### PR TITLE
Replace remaining uses of device_vector

### LINKED ISSUE #2631: [BUG] Mixing `rmm::device_vector` and non-default stream is unsafe
**Describe the bug**

As identified in https://github.com/rapidsai/cudf/issues/2630 mixing the use of an `rmm::device_vector` with kernels executing on the non-default stream is dangerous. 

For example,

```
rmm::device_vector<int> v(100); // allocation happens on null stream
some_kernel<<<...., non_default_stream>>>(v.data().get()...) // execution occurs on non-null stream
```
In non-pool mode, the above is safe because allocation defaults to `cudaMalloc`, which is synchronous.

However, if RMM is used with the pool enabled (i.e., using CNMEM), the above example is broken. `cnmemMalloc` is _asynchronous_ wrt to the host, and therefore `v` is not guaranteed to be allocated before `some_kernel` is launched.  

**Expected behavior**
There needs to be a safe way to mix use of a RAII object like `rmm::device_vector` with non-default streams. 

**Additional context**
RMM should provide a `device_vector` like abstraction that allows a way to specify a stream for allocation.

### PR DESCRIPTION
**Pull request by harrism**
Replaces `device_vector` with `device_uvector` in decomp_test.cu (now .cpp) and quantile.cu.

Also updates `DEVELOPER_GUIDE.md` to advise against using `device_vector`.

Closes #2631

### CODE REVIEW

**Comment by harrism** [comment]
rerun tests

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/tests/io/comp/decomp_test.cpp [hunk 1]
```diff
@@ -18,10 +18,13 @@
 
 #include <cudf_test/base_fixture.hpp>
 
-#include <vector>
+#include <cudf/utilities/span.hpp>
```
Comment (ttnghia) [inline cpp/tests/io/comp/decomp_test.cpp:21]: I don't see any span is used in this file?

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/io/comp/decomp_test.cpp (hunk 1 above)
Comment (harrism) [reply cpp/tests/io/comp/decomp_test.cpp:21]: Thanks. I had added spans but then removed.

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**